### PR TITLE
[Feat][mHC] Implement mHC kernels

### LINKED ISSUE #138: [New Op] manifold Hyper Connection (mHC)
## Operator Description

Manifold-Constrained Hyper-Connections (mHC) is a general framework that projects the residual connection space of hyper-connections onto a specific manifold to restore the identity mapping property, while incorporating rigorous infrastructure optimizations to ensure computational efficiency.
Inspired by both residual connections (RC) and hyper-connections (HC), mHC preserves HC’s key advantage over RC—namely, its ability to scale weights along the residual streaming dimension—while simultaneously mitigating the training instability commonly observed in HC-based architectures.

## Implementation Plan

Similar to residual connections, mHC is designed to work in conjunction with attention kernels (e.g., MLA) and feed-forward network (FFN) layers such as those based on Mixture of Experts (MoE).

### 1. Kernel Implementation (L1)

- [ ] **Kernel**: Implement TileLang kernel in `top/kernels/mHC/`
- [ ] **Verification**: Pass functional correctness tests

### 2. Op Definition (L2)

- [ ] **Interface**: Define `torch.ops` interface in `top/ops/mHC.py`
- [ ] **Unit Tests**: Implement `tests/test_mHC.py` (Compare vs PyTorch Ref)
  - [ ] FP16 (close: 1e-3)
  - [ ] BF16 (close: 1.6e-2)
- [ ] **Benchmarks**: Implement `benchmarks/benchmark_mHC.py`
  - [ ] Latency
  - [ ] TFLOPS
  - [ ] DRAM Bandwidth

### 3. Functional API (L3)

- [ ] **Autograd**: Implement `torch.autograd.Function` in `top/functions/mHC.py`
- [ ] **GradCheck**: Pass `torch.autograd.gradcheck`

### 4. Layer Wrapper (L4)

- [ ] **Module**: Implement `nn.Module` in `top/layers/mHC.py`
- [ ] **Type Hints**: Complete type hints (strict mode checked)

### 5. Benchmark Results

|       Shape        | FP16 Latency (ms) | FP16 TFLOPS | FP16 DRAM Bandwidth (GB/s) | BF16 Latency (ms) | BF16 TFLOPS | BF16 DRAM Bandwidth (GB/s) |
| :----------------: | :---------------: | :---------: | :------------------------: | :---------------: | :---------: | :------------------------: |
| (B=1, S=1024, ...) |        ...        |     ...     |            ...             |        ...        |     ...     |            ...             |

## Reference

[1] Z. Xie, Y. Wei, H. Cao, C. Zhao, C. Deng, J. Li, D. Dai, H. Gao, J. Chang, L. Zhao, S. Zhou, Z. Xu, Z. Zhang, W. Zeng, S. Hu, Y. Wang, J. Yuan, L. Wang, and W. Liang, "mHC: Manifold-Constrained Hyper-Connections," arXiv preprint arXiv:2512.24880, 2025. (https://arxiv.org/abs/2512.24880)

### PR DESCRIPTION
**Pull request by superAngGao**
resolve #138 

This PR adds the mHC kernel implementation and its necessary validation, improving decoder layer efficiency.

* Introduces the manifold-constrained hyper-connection (mHC) kernels with `mhc_pre` and `mhc_post` components for efficient processing in decoder layers.
* The `mhc_pre` kernel decomposes the input into attention and residual paths, while `mhc_post` fuses the outputs for the final layer result.
* Includes corresponding unit tests and benchmark scripts to validate functionality and performance.

### CODE REVIEW

File: benchmarks/mhc/mhc_post.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        self.batch = batch
+        self.n_expand = n_expand
+        self.c_x = c_x
+        self.dtype = dtype
+
+    @property
+    def total_flops(self) -> float:
+        flops = 2 * self.batch * (
+            self.n_expand * self.n_expand * self.c_x + self.n_expand * self.c_x)
+        return flops
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply benchmarks/mhc/mhc_post.py:23]: No, I'm correct this time.

File: benchmarks/mhc/mhc_pre.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+        self.n_expand = n_expand
+        self.c_x = c_x
+        self.dtype = dtype
+
+    @property
+    def total_flops(self) -> float:
+        flops = 2 * self.batch * (
+            (self.n_expand * self.n_expand * self.c_x * self.c_x) *
+            (self.n_expand * self.n_expand + 2 * self.n_expand) + self.n_expand * self.c_x)
+        return flops
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply benchmarks/mhc/mhc_pre.py:25]: No, I'm correct this time.

File: tests/ops/test_mhc_post.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        (2, 4, 1920, torch.bfloat16, False),
+        (4, 4, 2560, torch.bfloat16, False),
+    ],
+)
+def test_mhc_post_op(
+    batch: int,
+    n_expand: int,
+    c_x: int,
+    dtype: torch.dtype,
+    tune: bool,
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tests/ops/test_mhc_post.py:28]: Let's leave it here. More test parameters will be added when the autotune feature is ready.

File: tests/ops/test_mhc_pre.py [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        (2, 4, 1920, torch.float16, False),
+        (4, 4, 2560, torch.float16, False),
+    ],
+)
+def test_mhc_pre_op(
+    batch: int,
+    n_expand: int,
+    c_x: int,
+    dtype: torch.dtype,
+    tune: bool,
```
[the remark this answers is not in the document]
Comment (superAngGao) [reply tests/ops/test_mhc_pre.py:30]: Suggestion accepted. Thank you!

File: top/kernels/mhc/mhc_post.py [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+
+        result = _mhc_post_wrapped_kernel(self.batch, self.n_expand, self.c_x, self.dtype_str,
+                                          self.config["block_x_b"], self.config["block_C"],
+                                          self.config["num_stages"], self.config["threads"],
+                                          x_layer_out, h_post, x_res)
+        return result
+
+
+def main():
+    batch = 2
```
Comment (Renkeyiiiii) [inline top/kernels/mhc/mhc_post.py]: L1 layer Implementation should be raw tilelang kernels

File: top/kernels/mhc/mhc_pre.py [hunk 6]
```diff
@@ -0,0 +1,417 @@
...
+        result = _mhc_pre_wrapped_kernel(self.batch, self.n_expand, self.c_x, self.dtype_str,
+                                         self.config["block_x_b"], self.config["block_C"],
+                                         self.config["num_stages"], self.config["threads"], phi, x,
+                                         H, r, b, alpha_pre, alpha_post, alpha_res, H_pre, H_post,
+                                         H_res_0, sinkhorn_repeat, H_res, x_res)
+        return x_res, result
+
+
+def main():
+    batch = 4
```
Comment (Renkeyiiiii) [inline top/kernels/mhc/mhc_pre.py]: L1 layer Implementation should be raw tilelang kernels

File: top/kernels/mhc/mhc_post.py (hunk 5 above)
Comment (superAngGao) [reply top/kernels/mhc/mhc_post.py]: Thank you for your suggestion.

File: top/kernels/mhc/mhc_pre.py (hunk 6 above)
Comment (superAngGao) [reply top/kernels/mhc/mhc_pre.py]: Right. Thank you for your suggestion.
